src/wayland/seat/pointer.rs

```diff
@@ -0,0 +1,299 @@
+use std::sync::Mutex;
+
+use wayland_server::{
+    backend::{ClientId, ObjectId},
+    protocol::{
+        wl_pointer::{
+            self, Axis as WlAxis, AxisSource as WlAxisSource, ButtonState as WlButtonState, Request,
+            WlPointer,
+        },
+        wl_surface::WlSurface,
+    },
+    Dispatch, DisplayHandle, Resource,
+};
+
+use crate::{
+    backend::input::{Axis, AxisSource, ButtonState},
+    input::{
+        pointer::{
+            AxisFrame, ButtonEvent, CursorImageAttributes, CursorImageStatus, MotionEvent, PointerHandle,
+            PointerInternal, PointerTarget,
+        },
+        Seat,
+    },
+    utils::Serial,
+    wayland::compositor,
+};
+
+use super::{SeatHandler, SeatState, WaylandFocus};
+
+impl<D: SeatHandler> PointerHandle<D> {
+    pub(crate) fn new_pointer(&self, pointer: WlPointer) {
+        let mut guard = self.known_pointers.lock().unwrap();
+        guard.push(pointer);
+    }
+}
+
+/// WlSurface role of a cursor image icon
+pub const CURSOR_IMAGE_ROLE: &str = "cursor_image";
+
+fn with_focused_pointers<D: SeatHandler + 'static>(
+    seat: &Seat<D>,
+    surface: &WlSurface,
+    mut f: impl FnMut(WlPointer),
+) {
+    if let Some(pointer) = seat.get_pointer() {
+        let inner = pointer.known_pointers.lock().unwrap();
+        for ptr in &*inner {
+            if ptr.id().same_client_as(&surface.id()) {
+                f(ptr.clone())
+            }
+        }
+    }
+}
```

Comment:
Same here, I feel like `for_each_*` would be a better fit

Reply:
done